operators/src/adapters/raster_time_substream.rs

```diff
@@ -0,0 +1,266 @@
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use crate::util::Result;
+use futures::{
+    ready,
+    stream::{FusedStream, Stream},
+    Future,
+};
+use geoengine_datatypes::{
+    primitives::TimeInterval,
+    raster::{Pixel, RasterTile2D},
+};
+use pin_project::pin_project;
+
+#[must_use = "streams do nothing unless polled"]
+#[pin_project(project = RasterTimeSubStreamAdapterProjection)]
+pub struct RasterTimeMultiFold<St, P, A, AInit, F, Fut>
+where
+    St: Stream<Item = Result<RasterTile2D<P>>>,
+    P: Pixel,
+    AInit: FnMut() -> A,
+    F: FnMut(A, St::Item) -> Fut,
+    Fut: Future<Output = A>,
+{
+    #[pin]
+    stream: St,
+
+    accumulator_initializer: AInit,
+    f: F,
+
+    accum: Option<A>,
+    time: Option<TimeInterval>,
+
+    #[pin]
+    future: Option<Fut>,
+}
+
+impl<St, P, A, AInit, F, Fut> RasterTimeMultiFold<St, P, A, AInit, F, Fut>
+where
+    St: Stream<Item = Result<RasterTile2D<P>>>,
+    P: Pixel,
+    AInit: FnMut() -> A,
+    F: FnMut(A, St::Item) -> Fut,
+    Fut: Future<Output = A>,
+{
+    pub fn new(stream: St, mut accumulator_initializer: AInit, f: F) -> Self {
+        let accum = accumulator_initializer();
+
+        Self {
+            stream,
+            accumulator_initializer,
+            f,
+            accum: Some(accum),
+            future: None,
+            time: None,
+        }
+    }
+}
+
+impl<St, P, A, AInit, F, Fut> Stream for RasterTimeMultiFold<St, P, A, AInit, F, Fut>
+where
+    St: Stream<Item = Result<RasterTile2D<P>>>,
+    P: Pixel,
+    AInit: FnMut() -> A,
+    F: FnMut(A, St::Item) -> Fut,
+    Fut: Future<Output = A>,
+{
+    type Item = A;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.is_terminated() {
+            return Poll::Ready(None);
+        }
+
+        let mut this = self.project();
+
+        let value: A = loop {
```

Comment:
please add a note in which case the loop stops.

Reply:
ae24888